### PR TITLE
send admin a csv of donations every night

### PR DESCRIPTION
**Pull request by technillogue**

### CODE REVIEW

File: hotline/hotline.py [hunk 1]
```diff
@@ -122,6 +121,17 @@ async def talkback(self, msg: Message) -> Response:
         return None
 
 
+async def midnight_job(fn: Callable[[], Coroutine[Any, Any, None]]) -> None:
+    while 1:
+        now = datetime.datetime.now()
+        tomorrow = now + datetime.timedelta(days=1)
+        midnight = datetime.time.min
+        seconds_until_midnight = datetime.datetime.combine(tomorrow, midnight) - now
+        logging.info("sleeping %s seconds until midnight report")
```
Comment (transfaeries) [inline hotline/hotline.py:130]: you're not actually passing the seconds_until_midnight into logging should probably be:

logging.info("sleeping %s seconds until midnight report", seconds_until_midnight.total_seconds())

File: hotline/hotline.py [hunk 2]
```diff
@@ -510,6 +521,18 @@ async def do_give(self, msg: Message) -> Response:
...
+        header = "Donation UID, User UID, Timestamp, Amount MOB, Charity, FTXUSDPriceAtTimestamp"
+        body = [
+            f"{k}, {v}, {await self.mobster.get_historical_rate(v.split(', ')[1])}"
+            for (k, v) in self.donations.dict_.items()
+        ]
+        report_output = "\n".join([header, "\n"] + body)
+        report_filename = f"/tmp/HotlineDonations_{report_timestamp}.csv"
+        open(report_filename, "w").write(report_output)
+        await self.admin("Report Built", attachments=[report_filename])
+
```
Comment (transfaeries) [inline hotline/hotline.py:535]:
```suggestion
    @requires_admin
    async def do_report(self) -> None:
        """Generate donation report now as opposed to waiting till midnight"""
        await self.report()
```

Comment (transfaeries) [reply hotline/hotline.py:535]: might be neat to have a way to request one on the fly as opposed to just waiting till midnight

**Review by transfaeries — CHANGES_REQUESTED** [review]
I was able to test that it works by changing the seconds to midnight to be 60 seconds and it gave me my three reports in 3 minutes. Won't be able to know for sure without running overnight so we should do that somehow.

I also obviously, don't have any data so my csv only had headings, so please test with data also.

**Comment by itdaniher** [comment]
Probably only needs to be weekly? And maybe needs to be duplicated into Charity? We should also discuss deprecating this pipeline at some point...

**Comment by technillogue** [comment]
idk if there's another better way to do this now but i need it

**Pull request closed without merging**